crates/qasm2/src/bytecode.rs

```diff
@@ -0,0 +1,352 @@
+    #[pyo3(get)]
+    pub left: PyObject,
+    #[pyo3(get)]
+    pub right: PyObject,
+}
+
+/// Some custom callable Python function that the user told us about.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub struct ExprCustom {
+    #[pyo3(get)]
+    pub callable: PyObject,
+    #[pyo3(get)]
+    pub arguments: Vec<PyObject>,
+}
+
+/// Discriminator for the different types of unary operator.  We could have a separate class for
+/// each of these, but this way involves fewer imports in Python, and also serves to split up the
+/// option tree at the top level, so we don't have to test every unary operator before testing
+/// other operations.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub enum UnaryOpCode {
+    Negate,
+    Cos,
+    Exp,
+    Ln,
+    Sin,
+    Sqrt,
+    Tan,
+}
+
+/// Discriminator for the different types of binary operator.  We could have a separate class for
+/// each of these, but this way involves fewer imports in Python, and also serves to split up the
+/// option tree at the top level, so we don't have to test every binary operator before testing
+/// other operations.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub enum BinaryOpCode {
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Power,
+}
+
+/// An internal representation of the bytecode that will later be converted to the more free-form
+/// [Bytecode] Python-space objects.  This is fairly tightly coupled to Python space; the intent is
+/// just to communicate to Python as concisely as possible what it needs to do.  We want to have as
+/// little work to do in Python space as possible, since everything is slower there.
+///
+/// In various enumeration items, we use numeric keys to identify the object rather than its name.
+/// This is much more efficient in Python-space; rather than needing to build and lookup things in
+/// a hashmap, we can just build Python lists and index them directly, which also has the advantage
+/// of not needing to pass strings to Python for each gate.  It also gives us consistency with how
+/// qubits and clbits are tracked; there is no need to track both the register name and the index
+/// separately when we can use a simple single index.
+pub enum InternalBytecode {
+    Gate {
+        id: usize,
+        arguments: Vec<f64>,
+        qubits: Vec<usize>,
+    },
+    ConditionedGate {
+        id: usize,
+        arguments: Vec<f64>,
+        qubits: Vec<usize>,
+        creg: usize,
+        value: usize,
+    },
+    Measure {
+        qubit: usize,
+        clbit: usize,
+    },
+    ConditionedMeasure {
+        qubit: usize,
+        clbit: usize,
+        creg: usize,
+        value: usize,
+    },
+    Reset {
+        qubit: usize,
+    },
+    ConditionedReset {
+        qubit: usize,
+        creg: usize,
+        value: usize,
+    },
+    Barrier {
+        qubits: Vec<usize>,
+    },
+    DeclareQreg {
+        name: String,
+        size: usize,
```

Comment:
This is maybe an example of why I think type aliasing `object_key` would help. Is this `size` also an object key/index? Or it's the size you want the qreg to be?

Reply:
sure, it's a very valid point - I'm 90% sure I mixed up `start` and `size` in the broadcast code at least once when writing the parser.

(if it wasn't a rhetorical question, the answer is that it's the length of register, so it's a true `usize`)